### ISSUE TITLE
dsos_horn

### ISSUE THREAD

**Issue by blegat** [oxfordcontrol/COSMO.jl#96]
COSMO v0.5.0 is failing the `dsos_horn` test of SumOfSquares with:
```
dsos_horn: Error During Test at /home/blegat/.julia/dev/SumOfSquares/test/Tests/utilities.jl:47
  Got exception outside of a @test
  ArgumentError: column indices J[k] must satisfy 1 <= J[k] <= n
  Stacktrace:
   [1] sparse!(::Array{Int64,1}, ::Array{Int64,1}, ::Array{Float64,1}, ::Int64, ::Int64, ::typeof(+), ::Array{Int64,1}, ::Array{Int64,1}, ::Array{Int64,1}, ::Array{Float64,1}, ::Array{Int64,1}, ::Array{Int64,1}, ::Array{Float64,1}) at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.2/SparseArrays/src/sparsematrix.jl:613
   [2] sparse(::Array{Int64,1}, ::Array{Int64,1}, ::Array{Float64,1}, ::Int64, ::Int64, ::Function) at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.2/SparseArrays/src/sparsematrix.jl:525
   [3] sparse at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.2/SparseArrays/src/sparsematrix.jl:709 [inlined]
   [4] processconstraints(::Optimizer, ::MathOptInterface.Utilities.UniversalFallback{MathOptInterface.Utilities.Model{Float64}}, ::MathOptInterface.Utilities.IndexMap, ::Dict{Int64,UnitRange{Int64}}) at /home/blegat/.julia/packages/COSMO/Paq9I/src/MOIWrapper.jl:344
   [5] #copy_to#97(::Bool, ::typeof(MathOptInterface.copy_to), ::Optimizer, ::MathOptInterface.Utilities.UniversalFallback{MathOptInterface.Utilities.Model{Float64}}) at /home/blegat/.julia/packages/COSMO/Paq9I/src/MOIWrapper.jl:141
   [6] #copy_to at /home/blegat/.julia/dev/MathOptInterface/src/Utilities/cachingoptimizer.jl:0 [inlined]
   [7] attach_optimizer(::MathOptInterface.Utilities.CachingOptimizer{Optimizer,MathOptInterface.Utilities.UniversalFallback{MathOptInterface.Utilities.Model{Float64}}}) at /home/blegat/.julia/dev/MathOptInterface/src/Utilities/cachingoptimizer.jl:149
   [8] optimize!(::MathOptInterface.Utilities.CachingOptimizer{Optimizer,MathOptInterface.Utilities.UniversalFallback{MathOptInterface.Utilities.Model{Float64}}}) at /home/blegat/.julia/dev/MathOptInterface/src/Utilities/cachingoptimizer.jl:185
   [9] optimize!(::MathOptInterface.Bridges.LazyBridgeOptimizer{MathOptInterface.Utilities.CachingOptimizer{Optimizer,MathOptInterface.Utilities.UniversalFallback{MathOptInterface.Utilities.Model{Float64}}}}) at /home/blegat/.julia/dev/MathOptInterface/src/Bridges/bridge_optimizer.jl:225
   [10] optimize!(::MathOptInterface.Utilities.CachingOptimizer{MathOptInterface.AbstractOptimizer,MathOptInterface.Utilities.UniversalFallback{MathOptInterface.Utilities.Model{Float64}}}) at /home/blegat/.julia/dev/MathOptInterface/src/Utilities/cachingoptimizer.jl:189
   [11] #optimize!#78(::Bool, ::Bool, ::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}, ::typeof(optimize!), ::Model, ::Nothing) at /home/blegat/.julia/packages/JuMP/iGamg/src/optimizer_interface.jl:141
   [12] optimize! at /home/blegat/.julia/packages/JuMP/iGamg/src/optimizer_interface.jl:111 [inlined] (repeats 2 times)
   [13] horn_test(::OptimizerFactory, ::MathOptInterface.Test.TestConfig{Float64}, ::SumOfSquares.NonnegPolyInnerCone{SumOfSquares.DiagonallyDominantConeTriangle}) at /home/blegat/.julia/dev/SumOfSquares/test/Tests/horn.jl:40
   [14] dsos_horn_test(::OptimizerFactory, ::MathOptInterface.Test.TestConfig{Float64}) at /home/blegat/.julia/dev/SumOfSquares/test/Tests/horn.jl:67
   [15] macro expansion at /home/blegat/.julia/dev/SumOfSquares/test/Tests/utilities.jl:48 [inlined]
   [16] macro expansion at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.2/Test/src/Test.jl:1113 [inlined]
   [17] linear_test(::OptimizerFactory, ::MathOptInterface.Test.TestConfig{Float64}, ::Array{String,1}) at /home/blegat/.julia/dev/SumOfSquares/test/Tests/utilities.jl:48
   [18] linear_test(::OptimizerFactory, ::MathOptInterface.Test.TestConfig{Float64}) at /home/blegat/.julia/dev/SumOfSquares/test/Tests/utilities.jl:43
   [19] top-level scope at /home/blegat/.julia/dev/SumOfSquares/test/Solvers/cosmo_tests.jl:6
   [20] top-level scope at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.2/Test/src/Test.jl:1113
   [21] top-level scope at /home/blegat/.julia/dev/SumOfSquares/test/Solvers/cosmo_tests.jl:6
   [22] include at ./boot.jl:328 [inlined]
   [23] include_relative(::Module, ::String) at ./loading.jl:1094
   [24] include(::Module, ::String) at ./Base.jl:31
   [25] include(::String) at ./client.jl:431
   [26] top-level scope at REPL[42]:1
   [27] eval(::Module, ::Any) at ./boot.jl:330
   [28] eval_user_input(::Any, ::REPL.REPLBackend) at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.2/REPL/src/REPL.jl:86
   [29] macro expansion at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.2/REPL/src/REPL.jl:118 [inlined]
   [30] (::getfield(REPL, Symbol("##26#27")){REPL.REPLBackend})() at ./task.jl:268
```
See https://github.com/JuliaOpt/SumOfSquares.jl/blob/master/test/Tests/horn.jl

**Comment by migarstka** [comment]
Thanks Benoît. 

This is strange behaviour. It seems to be an issue with `delete(model, cref)` or the way I reset the COSMO-model is reset before the 2nd problem is solved. If you run the three problems in that script individually you get the expected result (without errors)

**Comment by blegat** [comment]
I think the only difference is that in the second case, the model will be emptied by the CachingOptimizer (instead of having a fresh new one) and the cache that will be copied to COSMO will have different indices (since it will have applied the deletion).

> Both examples yield very different IndexMaps.

So the IndexMaps will be different indeed

**Comment by migarstka** [comment]
Fixed in f90cc62

The problem was that I didn't map the variable indices involved in the constraints through the `idxmap`. I am surprised that this case isn't covered anywhere in the MOI unit tests. We should probably add an appropriate test for this.
